### PR TITLE
Support RawValue with [AppDomainFixedOutput]

### LINKED ISSUE #3992: Don't combine IRawValue with [AppDomainFixed]
From @snakefoot 
> IRawValue doesn't work in combination with [AppDomainFixedOutput] (Will always be turned into a literal).
> 
> Actually there should be made a unit-test like this:
> 
>         [Fact]
>         public void IRawValueRenderer_AppDomainFixedOutput_Attribute_NotRequired()
>         {
>             foreach (Type type in allTypes)
>             {
>                 if (typeof(NLog.Internal.IRawValue).IsAssignableFrom(type) && !type.IsInterface)
>                 {
>                     var appDomainFixedOutputAttribute = type.GetCustomAttribute<AppDomainFixedOutputAttribute>();
>                     Assert.True(ReferenceEquals(appDomainFixedOutputAttribute, null), $"{type.ToString()} should not implement IStringValueRenderer");
>                 }
>             }
>         }
>

### PR DESCRIPTION
**Pull request by 304NotModified**
- plus some small code improvements in the file (Boy Scout Rule) 

Fixes #3992, alternative to https://github.com/NLog/NLog/pull/3993

### CODE REVIEW

**Comment by snakefoot** [comment]
Think this will break the whole literal optimization as merging stops working along with SimpleLayout fixed text.


Think you should just extend the existing LiteralLayoutRenderer and when merging two into one then reset raw value and mark it as raw value not available.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified<https://github.com/304NotModified> requested your review on: #3997<https://github.com/NLog/NLog/pull/3997> Improve ConvertToLiteral to handle RawValues.

**Comment by 304NotModified** [comment]
Good one indeed!

**Comment by 304NotModified** [comment]
@snakefoot all fixed now, please check :)

File: src/NLog/Layouts/LayoutParser.cs [hunk 1]
```diff
@@ -494,13 +499,11 @@ private static bool CanBeConvertedToLiteral(LayoutRenderer lr)
...
+        private static void MergeLiterals(IList<LayoutRenderer> list)
         {
             for (int i = 0; i + 1 < list.Count;)
             {
-                var lr1 = list[i] as LiteralLayoutRenderer;
-                var lr2 = list[i + 1] as LiteralLayoutRenderer;
-                if (lr1 != null && lr2 != null)
+                if (list[i] is LiteralLayoutRenderer lr1 && list[i + 1] is LiteralLayoutRenderer lr2)
                 {
                     lr1.Text += lr2.Text;
```
Comment (snakefoot) [inline src/NLog/Layouts/LayoutParser.cs:508]: You forgot to reset the ability to return raw-value when merging two literals into one.

Comment (304NotModified) [reply src/NLog/Layouts/LayoutParser.cs:508]: Well when combining two you haven't a raw value? 

Dunno which test case I'm missing then. (I solved this TDD style)

Comment (snakefoot) [reply src/NLog/Layouts/LayoutParser.cs:508]: What raw value does this Layout return: `${processid}-${processname}` ?

Comment (304NotModified) [reply src/NLog/Layouts/LayoutParser.cs:508]: Fixed!

File: src/NLog/Layouts/LayoutParser.cs [hunk 2]
```diff
@@ -512,9 +521,17 @@ private static void MergeLiterals(List<LayoutRenderer> list)
             }
         }
 
-        private static LayoutRenderer ConvertToLiteral(LayoutRenderer renderer)
+        private static LayoutRenderer ConvertToLiteral(IRenderable renderer)
```
Comment (snakefoot) [inline src/NLog/Layouts/LayoutParser.cs:524]: Think it should stay as type `LayoutRenderer`. Less confusing.

File: src/NLog/Layouts/LayoutParser.cs [hunk 3]
```diff
@@ -512,9 +521,17 @@ private static void MergeLiterals(List<LayoutRenderer> list)
...
         {
-            return new LiteralLayoutRenderer(renderer.Render(LogEventInfo.CreateNullEvent()));
+            var logEventInfo = LogEventInfo.CreateNullEvent();
+            var text = renderer.Render(logEventInfo);
+            if (renderer is IRawValue rawValueRender)
+            {
+                var success = rawValueRender.TryGetRawValue(logEventInfo, out var rawValue);
```
Comment (snakefoot) [inline src/NLog/Layouts/LayoutParser.cs:530]: When not a success, then no need to change into `LiteralWithRawValueLayoutRenderer`

Comment (304NotModified) [reply src/NLog/Layouts/LayoutParser.cs:530]: Well less tricky to keep it like this.

Comment (304NotModified) [reply src/NLog/Layouts/LayoutParser.cs:530]: close as this is by design

File: src/NLog/Layouts/LayoutParser.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Layouts/LayoutParser.cs]: not really an opinion on this, so keep it LayoutRenderer

**Comment by 304NotModified** [comment]
> But looking good. Could consider re-base to master for ver. 4.7.3, since the change is actually very discrete.

I think we should only do bug fixes on 4.7.x for now

**Pull request merged**